api/handler.go

```diff
@@ -58,145 +52,62 @@ func RegisterHandlers(m manager.Manager) error {
 	return nil
 }
 
-func handleRequest(m manager.Manager, w http.ResponseWriter, r *http.Request) error {
+// Captures the API version, requestType [optional], and remaining request [optional].
+var apiRegexp = regexp.MustCompile("/api/([^/]+)/?([^/]+)?(.*)")
+
+func handleRequest(supportedApiVersions map[string]ApiVersion, m manager.Manager, w http.ResponseWriter, r *http.Request) error {
 	start := time.Now()
-	defer glog.V(2).Infof("Request took %s", time.Since(start))
+	defer func() {
+		glog.V(2).Infof("Request took %s", time.Since(start))
+	}()
 
 	request := r.URL.Path
-	requestElements := strings.Split(r.URL.Path, "/")
 
-	// Verify that we have all the elements we expect:
-	// <empty>/api/<version>/<request type>[/<args...>]
-	// [0]     [1] [2]       [3]             [4...]
-	if len(requestElements) < 4 {
+	const apiPrefix = "/api"
+	if !strings.HasPrefix(request, apiPrefix) {
 		return fmt.Errorf("incomplete API request %q", request)
 	}
 
-	// Get all the element parts.
-	emptyElement := requestElements[0]
-	apiElement := requestElements[1]
-	version := requestElements[2]
-	requestType := requestElements[3]
-	requestArgs := []string{}
-	if len(requestElements) > 4 {
-		requestArgs = requestElements[4:]
+	// If the request doesn't have an API version, list those.
+	if request == apiPrefix || request == apiResource {
+		versions := make([]string, 0, len(supportedApiVersions))
+		for v := range supportedApiVersions {
+			versions = append(versions, v)
+		}
+		sort.Strings(versions)
+		fmt.Fprintf(w, "Supported API versions: %s", strings.Join(versions, ","))
+		return nil
 	}
 
-	// The container name is the path after the requestType.
-	containerName := path.Join("/", strings.Join(requestArgs, "/"))
-
-	// Check elements.
-	if len(emptyElement) != 0 {
-		return fmt.Errorf("unexpected API request format %q", request)
-	}
-	if apiElement != "api" {
-		return fmt.Errorf("invalid API request format %q", request)
+	// Verify that we have all the elements we expect:
+	// /<version>/<request type>[/<args...>]
+	requestElements := apiRegexp.FindStringSubmatch(request)
+	if len(requestElements) == 0 {
+		return fmt.Errorf("malformed request %q", request)
 	}
-	if _, ok := supportedApiVersions[version]; !ok {
+	version := requestElements[1]
```

Comment:
I was referring to something like the following:
`const (
apiVersion = iota + 1
apiRequestType
apiRequestArgs
)`, instead of using numeric indexes.
Sorry if that was not clear in the previous comment.